From 2fde00da6f73a6271a7bb60f9716f642cb0312dc Mon Sep 17 00:00:00 2001
From: bill <84451962+bill997@users.noreply.github.com>
Date: Thu, 1 May 2025 00:50:58 +1000
Subject: [PATCH] Fix compatibility with protobuf v30 (cpp 6.30.0) - similar to
 gz-msgs #499 (#615)

* Fix compatibility with protobuf v30 (cpp 6.30.0) - similar to gz-msgs #499

Signed-off-by: bill ross <4451962+bill997@users.noreply.github.com>
---
 include/gz/transport/RepHandler.hh            |  4 ++--
 include/gz/transport/ReqHandler.hh            |  8 ++++----
 include/gz/transport/SubscriptionHandler.hh   |  2 +-
 include/gz/transport/detail/Node.hh           | 19 ++++++++++++-------
 parameters/src/Client.cc                      |  2 +-
 parameters/src/Registry.cc                    | 13 +++++++------
 src/CIface_TEST.cc                            | 10 +++++-----
 src/HandlerStorage_TEST.cc                    |  7 ++++---
 src/Node.cc                                   |  2 +-
 src/Node_TEST.cc                              | 19 ++++++++++++-------
 src/cmd/gz.cc                                 |  2 +-
 src/cmd/gz_TEST.cc                            |  3 ++-
 .../twoProcsPubSubSingleSubscriber_aux.cc     |  2 +-
 .../twoProcsPubSubSubscriber_aux.cc           |  2 +-
 test/integration/twoProcsPubSub.cc            |  7 ++++---
 15 files changed, 58 insertions(+), 44 deletions(-)

diff --git include/gz/transport/RepHandler.hh include/gz/transport/RepHandler.hh
index b8406439b..ca1c578de 100644
--- include/gz/transport/RepHandler.hh
+++ include/gz/transport/RepHandler.hh
@@ -237,13 +237,13 @@ namespace gz
       // Documentation inherited.
       public: virtual std::string ReqTypeName() const
       {
-        return Req().GetTypeName();
+        return std::string(Req().GetTypeName());
       }
 
       // Documentation inherited.
       public: virtual std::string RepTypeName() const
       {
-        return Rep().GetTypeName();
+        return std::string(Rep().GetTypeName());
       }
 
       /// \brief Create a specific protobuf message given its serialized data.
diff --git include/gz/transport/ReqHandler.hh include/gz/transport/ReqHandler.hh
index cea0664dd..df3404581 100644
--- include/gz/transport/ReqHandler.hh
+++ include/gz/transport/ReqHandler.hh
@@ -285,13 +285,13 @@ namespace gz
       // Documentation inherited.
       public: virtual std::string ReqTypeName() const
       {
-        return Req().GetTypeName();
+        return std::string(Req().GetTypeName());
       }
 
       // Documentation inherited.
       public: virtual std::string RepTypeName() const
       {
-        return Rep().GetTypeName();
+        return std::string(Rep().GetTypeName());
       }
 
       /// \brief Protobuf message containing the request's parameters.
@@ -382,7 +382,7 @@ namespace gz
       public: virtual std::string ReqTypeName() const
       {
         if (this->reqMsg)
-          return this->reqMsg->GetTypeName();
+          return std::string(this->reqMsg->GetTypeName());
         else
         {
           std::cerr << "ReqHandler::ReqTypeName() Warning: Using ReqTypeName() "
@@ -395,7 +395,7 @@ namespace gz
       public: virtual std::string RepTypeName() const
       {
         if (this->repMsg)
-          return this->repMsg->GetTypeName();
+          return std::string(this->repMsg->GetTypeName());
         else
         {
           std::cerr << "ReqHandler::RepTypeName() Warning: Using RepTypeName() "
diff --git include/gz/transport/SubscriptionHandler.hh include/gz/transport/SubscriptionHandler.hh
index ed705ad92..a612cc40d 100644
--- include/gz/transport/SubscriptionHandler.hh
+++ include/gz/transport/SubscriptionHandler.hh
@@ -188,7 +188,7 @@ namespace gz
       // Documentation inherited.
       public: std::string TypeName()
       {
-        return T().GetTypeName();
+        return std::string(T().GetTypeName());
       }
 
       /// \brief Set the callback for this handler.
diff --git include/gz/transport/detail/Node.hh include/gz/transport/detail/Node.hh
index d069ebdbe..2815e9dee 100644
--- include/gz/transport/detail/Node.hh
+++ include/gz/transport/detail/Node.hh
@@ -34,7 +34,8 @@ namespace gz
         const std::string &_topic,
         const AdvertiseMessageOptions &_options)
     {
-      return this->Advertise(_topic, MessageT().GetTypeName(), _options);
+      return this->Advertise(_topic,
+          std::string(MessageT().GetTypeName()), _options);
     }
 
     //////////////////////////////////////////////////
@@ -267,7 +268,8 @@ namespace gz
         this->Shared()->myReplierAddress,
         this->Shared()->replierId.ToString(),
         this->Shared()->pUuid, this->NodeUuid(),
-        RequestT().GetTypeName(), ReplyT().GetTypeName(), _options);
+        std::string(RequestT().GetTypeName()),
+        std::string(ReplyT().GetTypeName()), _options);
 
       if (!this->Shared()->AdvertisePublisher(publisher))
       {
@@ -420,8 +422,8 @@ namespace gz
         std::lock_guard<std::recursive_mutex> lk(this->Shared()->mutex);
         localResponserFound = this->Shared()->repliers.FirstHandler(
               fullyQualifiedTopic,
-              RequestT().GetTypeName(),
-              ReplyT().GetTypeName(),
+              std::string(RequestT().GetTypeName()),
+              std::string(ReplyT().GetTypeName()),
               repHandler);
       }
 
@@ -458,7 +460,8 @@ namespace gz
         if (this->Shared()->TopicPublishers(fullyQualifiedTopic, addresses))
         {
           this->Shared()->SendPendingRemoteReqs(fullyQualifiedTopic,
-            RequestT().GetTypeName(), ReplyT().GetTypeName());
+            std::string(RequestT().GetTypeName()),
+            std::string(ReplyT().GetTypeName()));
         }
         else
         {
@@ -551,7 +554,8 @@ namespace gz
       // If the responser is within my process.
       IRepHandlerPtr repHandler;
       if (this->Shared()->repliers.FirstHandler(fullyQualifiedTopic,
-        _request.GetTypeName(), _reply.GetTypeName(), repHandler))
+        std::string(_request.GetTypeName()),
+        std::string(_reply.GetTypeName()), repHandler))
       {
         // There is a responser in my process, let's use it.
         _result = repHandler->RunLocalCallback(_request, _reply);
@@ -567,7 +571,8 @@ namespace gz
       if (this->Shared()->TopicPublishers(fullyQualifiedTopic, addresses))
       {
         this->Shared()->SendPendingRemoteReqs(fullyQualifiedTopic,
-          _request.GetTypeName(), _reply.GetTypeName());
+          std::string(_request.GetTypeName()),
+          std::string(_reply.GetTypeName()));
       }
       else
       {
diff --git parameters/src/Client.cc parameters/src/Client.cc
index 02aaeb219..a932ae687 100644
--- parameters/src/Client.cc
+++ parameters/src/Client.cc
@@ -216,7 +216,7 @@ ParametersClient::DeclareParameter(
     return ParameterResult{
       ParameterResultType::InvalidType,
       _parameterName,
-      _msg.GetDescriptor()->name()};
+      std::string(_msg.GetDescriptor()->name())};
   }
   return ParameterResult{ParameterResultType::Unexpected, _parameterName};
 }
diff --git parameters/src/Registry.cc parameters/src/Registry.cc
index 0e6d308b3..4a5f57fcf 100644
--- parameters/src/Registry.cc
+++ parameters/src/Registry.cc
@@ -144,7 +144,7 @@ bool ParametersRegistryPrivate::ListParameters(const msgs::Empty &,
       auto * decl = _res.add_parameter_declarations();
       decl->set_name(paramPair.first);
       decl->set_type(addGzMsgsPrefix(
-        paramPair.second->GetDescriptor()->name()));
+        std::string(paramPair.second->GetDescriptor()->name())));
     }
   }
   return true;
@@ -238,7 +238,7 @@ ParametersRegistry::DeclareParameter(
   const std::string & _parameterName,
   const google::protobuf::Message & _msg)
 {
-  auto protoType = addGzMsgsPrefix(_msg.GetDescriptor()->name());
+  auto protoType = addGzMsgsPrefix(std::string(_msg.GetDescriptor()->name()));
   auto newParam = gz::msgs::Factory::New(protoType);
   if (!newParam) {
     return ParameterResult{
@@ -270,7 +270,7 @@ ParametersRegistry::Parameter(
     return ParameterResult{
       ParameterResultType::InvalidType,
       _parameterName,
-      addGzMsgsPrefix(protoType)};
+      addGzMsgsPrefix(std::string(protoType))};
   }
   _parameter.CopyFrom(*it->second);
   return ParameterResult{ParameterResultType::Success};
@@ -290,12 +290,12 @@ ParametersRegistry::Parameter(
       _parameterName};
   }
   const auto & protoType = it->second->GetDescriptor()->name();
-  _parameter = gz::msgs::Factory::New(protoType);
+  _parameter = gz::msgs::Factory::New(std::string(protoType));
   if (!_parameter) {
     return ParameterResult{
       ParameterResultType::InvalidType,
       _parameterName,
-      addGzMsgsPrefix(protoType)};
+      addGzMsgsPrefix(std::string(protoType))};
 
   }
   _parameter->CopyFrom(*it->second);
@@ -320,7 +320,8 @@ ParametersRegistry::SetParameter(
     return ParameterResult{
       ParameterResultType::InvalidType,
       _parameterName,
-      addGzMsgsPrefix(it->second->GetDescriptor()->name())};
+      std::string(addGzMsgsPrefix(
+            std::string(it->second->GetDescriptor()->name())))};
   }
   it->second = std::move(_value);
   return ParameterResult{ParameterResultType::Success};
diff --git src/CIface_TEST.cc src/CIface_TEST.cc
index b67f3c979..55b4e0e61 100644
--- src/CIface_TEST.cc
+++ src/CIface_TEST.cc
@@ -96,7 +96,7 @@ TEST(CIfaceTest, PubSub)
   msg.SerializeToArray(buffer, size);
 
   EXPECT_EQ(0,
-    gzTransportPublish(node, topic, buffer, msg.GetTypeName().c_str()));
+    gzTransportPublish(node, topic, buffer, msg.GetTypeName().data()));
 
   EXPECT_EQ(2, count);
 
@@ -105,7 +105,7 @@ TEST(CIfaceTest, PubSub)
   // Unsubscribe
   ASSERT_EQ(0, gzTransportUnsubscribe(node, topic));
   EXPECT_EQ(0,
-    gzTransportPublish(node, topic, buffer, msg.GetTypeName().c_str()));
+    gzTransportPublish(node, topic, buffer, msg.GetTypeName().data()));
   EXPECT_EQ(0, count);
 
   free(buffer);
@@ -153,12 +153,12 @@ TEST(CIfaceTest, PubSubPartitions)
 
   // Publish on "bar" partition
   EXPECT_EQ(0,
-    gzTransportPublish(nodeBar, topic, buffer, msg.GetTypeName().c_str()));
+    gzTransportPublish(nodeBar, topic, buffer, msg.GetTypeName().data()));
   EXPECT_EQ(1, count);
 
   // Publish on default partition
   EXPECT_EQ(0,
-    gzTransportPublish(nodeBar, topic, buffer, msg.GetTypeName().c_str()));
+    gzTransportPublish(nodeBar, topic, buffer, msg.GetTypeName().data()));
   EXPECT_EQ(2, count);
 
   count = 0;
@@ -166,7 +166,7 @@ TEST(CIfaceTest, PubSubPartitions)
   // Unsubscribe
   ASSERT_EQ(0, gzTransportUnsubscribe(nodeBar, topic));
   EXPECT_EQ(0,
-    gzTransportPublish(nodeBar, topic, buffer, msg.GetTypeName().c_str()));
+    gzTransportPublish(nodeBar, topic, buffer, msg.GetTypeName().data()));
   EXPECT_EQ(0, count);
 
   free(buffer);
diff --git src/HandlerStorage_TEST.cc src/HandlerStorage_TEST.cc
index fa2d107bc..638f5a70d 100644
--- src/HandlerStorage_TEST.cc
+++ src/HandlerStorage_TEST.cc
@@ -67,8 +67,8 @@ TEST(RepStorageTest, RepStorageAPI)
   msgs::Int32 rep1Msg;
   bool result;
   msgs::Vector3d reqMsg;
-  std::string reqType = reqMsg.GetTypeName();
-  std::string rep1Type = rep1Msg.GetTypeName();
+  std::string reqType = std::string(reqMsg.GetTypeName());
+  std::string rep1Type = std::string(rep1Msg.GetTypeName());
 
   reqMsg.set_x(1.0);
   reqMsg.set_y(2.0);
@@ -234,7 +234,8 @@ TEST(RepStorageTest, SubStorageNoCallbacks)
   EXPECT_FALSE(subs.FirstHandler(topic, "incorrect type", handler));
 
   // Now try to retrieve the first callback with the correct type.
-  EXPECT_TRUE(subs.FirstHandler(topic, msg.GetTypeName(), handler));
+  EXPECT_TRUE(subs.FirstHandler(topic,
+        std::string(msg.GetTypeName()), handler));
 
   // Verify the handler.
   EXPECT_EQ(handler->TypeName(), sub1HandlerPtr->TypeName());
diff --git src/Node.cc src/Node.cc
index 7e347921d..0e7b19369 100644
--- src/Node.cc
+++ src/Node.cc
@@ -420,7 +420,7 @@ bool Node::Publisher::Publish(const ProtoMsg &_msg)
     };
 
     if (!this->dataPtr->shared->Publish(this->dataPtr->publisher.Topic(),
-          msgBuffer, msgSize, myDeallocator, _msg.GetTypeName()))
+          msgBuffer, msgSize, myDeallocator, std::string(_msg.GetTypeName())))
     {
       return false;
     }
diff --git src/Node_TEST.cc src/Node_TEST.cc
index 2df093614..cb3bd1d88 100644
--- src/Node_TEST.cc
+++ src/Node_TEST.cc
@@ -346,8 +346,9 @@ class MyTestClass
     // Request a valid service using RequestRaw.
     std::string reqStr, repStr, repTypeName;
     req.SerializeToString(&reqStr);
-    EXPECT_TRUE(this->node.RequestRaw(g_topic, reqStr, req.GetTypeName(),
-          rep.GetTypeName(), timeout, repStr, result));
+    EXPECT_TRUE(this->node.RequestRaw(g_topic, reqStr,
+          std::string(req.GetTypeName()),
+          std::string(rep.GetTypeName()), timeout, repStr, result));
     rep.ParseFromString(repStr);
     ASSERT_TRUE(result);
     EXPECT_EQ(rep.data(), data);
@@ -558,7 +559,7 @@ TEST(NodeTest, PubWithoutAdvertise)
   EXPECT_TRUE(node1.SubscribedTopics().empty());
   EXPECT_TRUE(node1.AdvertisedServices().empty());
 
-  auto pub1 = node1.Advertise(g_topic, msg.GetTypeName());
+  auto pub1 = node1.Advertise(g_topic, std::string(msg.GetTypeName()));
   EXPECT_TRUE(pub1);
 
   auto advertisedTopics = node1.AdvertisedTopics();
@@ -747,7 +748,8 @@ TEST(NodeTest, RawPubSubSameThreadMessageInfo)
   std::this_thread::sleep_for(std::chrono::milliseconds(100));
 
   // Publish a first message.
-  EXPECT_TRUE(pub.PublishRaw(msg.SerializeAsString(), msg.GetTypeName()));
+  EXPECT_TRUE(pub.PublishRaw(msg.SerializeAsString(),
+        std::string(msg.GetTypeName())));
 
   // Give some time to the subscribers.
   std::this_thread::sleep_for(std::chrono::milliseconds(100));
@@ -758,7 +760,8 @@ TEST(NodeTest, RawPubSubSameThreadMessageInfo)
   reset();
 
   // Publish a second message on topic.
-  EXPECT_TRUE(pub.PublishRaw(msg.SerializeAsString(), msg.GetTypeName()));
+  EXPECT_TRUE(pub.PublishRaw(msg.SerializeAsString(),
+        std::string(msg.GetTypeName())));
 
   // Give some time to the subscribers.
   std::this_thread::sleep_for(std::chrono::milliseconds(100));
@@ -787,7 +790,8 @@ TEST(NodeTest, RawPubRawSubSameThreadMessageInfo)
   std::this_thread::sleep_for(std::chrono::milliseconds(100));
 
   // Publish a first message.
-  EXPECT_TRUE(pub.PublishRaw(msg.SerializeAsString(), msg.GetTypeName()));
+  EXPECT_TRUE(pub.PublishRaw(msg.SerializeAsString(),
+        std::string(msg.GetTypeName())));
 
   // Give some time to the subscribers.
   std::this_thread::sleep_for(std::chrono::milliseconds(100));
@@ -798,7 +802,8 @@ TEST(NodeTest, RawPubRawSubSameThreadMessageInfo)
   reset();
 
   // Publish a second message on topic.
-  EXPECT_TRUE(pub.PublishRaw(msg.SerializeAsString(), msg.GetTypeName()));
+  EXPECT_TRUE(pub.PublishRaw(msg.SerializeAsString(),
+        std::string(msg.GetTypeName())));
 
   // Give some time to the subscribers.
   std::this_thread::sleep_for(std::chrono::milliseconds(100));
diff --git src/cmd/gz.cc src/cmd/gz.cc
index adb948c16..fc98d86f9 100644
--- src/cmd/gz.cc
+++ src/cmd/gz.cc
@@ -186,7 +186,7 @@ extern "C" void cmdTopicPub(const char *_topic,
   {
     // Create the node and advertise the topic
     Node node;
-    auto pub = node.Advertise(_topic, msg->GetTypeName());
+    auto pub = node.Advertise(_topic, std::string(msg->GetTypeName()));
 
     // Publish the message
     if (pub)
diff --git src/cmd/gz_TEST.cc src/cmd/gz_TEST.cc
index 4050e5840..143d1819e 100644
--- src/cmd/gz_TEST.cc
+++ src/cmd/gz_TEST.cc
@@ -142,7 +142,8 @@ TEST(gzTest, TopicListSub)
   transport::Node node;
   node.Subscribe("/foo", topicCB);
   node.Subscribe("/bar", genericCb);
-  node.SubscribeRaw("/baz", cbRaw, msgs::StringMsg().GetTypeName());
+  node.SubscribeRaw("/baz", cbRaw,
+      std::string(msgs::StringMsg().GetTypeName()));
   node.Subscribe("/no", topicCB);
   node.Unsubscribe("/no");
 
